lib/cmd_grepdom.go

```diff
@@ -0,0 +1,221 @@
+package lib
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/fatih/color"
+	"github.com/spf13/pflag"
+	"golang.org/x/net/idna"
+)
+
+// regular expression that matches simple domains and IDN ones.
+var domainRegex = regexp.MustCompile(`\b(?:[a-zA-Z0-9-]+\.){1,}[a-zA-Z]{2,}\b|([^\s.]+\.)+[^\s]{2,}\b`)
+
+type CmdGrepDomainFlags struct {
+	OnlyMatching bool
+	NoFilename   bool
+	NoRecurse    bool
+	Help         bool
+	NoColor      bool
+	ExcludePuny  bool
+}
+
+func (f *CmdGrepDomainFlags) Init() {
+	pflag.BoolVarP(
+		&f.OnlyMatching,
+		"only-matching", "o", false,
+		"print only matched domains in result line.",
+	)
+	pflag.BoolVarP(
+		&f.NoFilename,
+		"no-filename", "h", false,
+		"don't print source of match in result lines when more than 1 source.",
+	)
+	pflag.BoolVarP(
+		&f.NoRecurse,
+		"no-recurse", "", false,
+		"don't recurse into more dirs in dir sources.",
+	)
+	pflag.BoolVarP(
+		&f.Help,
+		"help", "", false,
+		"show help.",
+	)
+	pflag.BoolVarP(
+		&f.NoColor,
+		"nocolor", "", false,
+		"disable color output.",
+	)
+	pflag.BoolVarP(
+		&f.ExcludePuny,
+		"no-punycode", "n", false,
+		"do not convert domains to punycode.",
+	)
+}
+
+// CmdGrepDomain is the common core logic for the grepdom command.
+func CmdGrepDomain(f CmdGrepDomainFlags, args []string, printHelp func()) error {
+	if f.NoColor {
+		color.NoColor = true
+	}
+
+	if f.Help {
+		printHelp()
+		return nil
+	}
+
+	// require args.
+	stat, _ := os.Stdin.Stat()
+	isStdin := (stat.Mode() & os.ModeCharDevice) == 0
+	srcCnt := len(args)
+	if isStdin {
+		srcCnt += 1
+	}
+	if srcCnt == 0 {
+		printHelp()
+		return nil
+	}
+
+	// if user hasn't forced no-filename, and we have more than 1 source, then
+	// output file
+	if !f.NoFilename && srcCnt > 1 {
+		f.NoFilename = false
+	} else {
+		f.NoFilename = true
+	}
+
+	fmtSrc := color.New(color.FgMagenta)
+	fmtMatch := color.New(color.Bold, color.FgRed)
+
+	rexp := domainRegex
+
+	// Actual scanner logic for domains
+	scanrdr := func(src string, r io.Reader) {
+		var hitEOF bool
+		buf := bufio.NewReader(r)
+
+		for {
+			if hitEOF {
+				return
+			}
+
+			d, err := buf.ReadString('\n')
+			if err == io.EOF {
+				if len(d) == 0 {
+					return
+				}
+
+				// do one more loop for remaining content
+				hitEOF = true
+			} else if err != nil {
+				return
+			}
+
+			var matches [][]int
+			allMatches := rexp.FindAllStringIndex(d, -1)
```

Comment:
If they enabled punycode, we need to punycode-ify the line we're reading. E.g. imagine what would happen if this line `d` currently has a chinese character domain. Would it work against your regexp?

Reply:
I have currently tested the regex on German and Japanese character domains and it matched fine. I'll test Chinese character domains too. Also, I wrongly assumed that if they enable punycode, we have to print line `d` that highlights matched domains and then print that domain along with its punycode, separately.

So if I'm understanding correctly, we need to print line `d` that replaces the original domain with the punycode one if they enable it?

Reply:
Hmm I'm surprised it does though because the regex isnt supporting non-ASCII chars is it?

We can just print the actual line as it was, with the relevant match highlighted, whether the match is in punycode or not. We'll likely need to test some cases and see samples to see if this behavior is what is desired. New stuff here.

Reply:
You're right. That regex only matched domains that were either entirely ASCII or a mix between ASCII and non-ASCII chars but not entirely non-ASCII chars. I've tried altering it to accept those, too.

Attached are the two screenshots. One shows a file's contents to be parsed by `grepdom`. The other shows the output in the terminal.

![file](https://github.com/ipinfo/cli/assets/55349478/810e4002-0052-4d72-a06e-5cfdbca5c4b9)


![terminal](https://github.com/ipinfo/cli/assets/55349478/9ab09a9f-d179-4f35-9376-18622a3170cd)

Reply:
That's better